cli/lsp/documents.rs

```diff
@@ -1248,21 +1242,108 @@ impl Documents {
 
     // only refresh the dependencies if the underlying configuration has changed
     if self.resolver_config_hash != new_resolver_config_hash {
-      self.refresh_dependencies();
+      self.refresh_dependencies(root_dirs);
       self.resolver_config_hash = new_resolver_config_hash;
     }
 
     self.dirty = true;
   }
 
-  fn refresh_dependencies(&mut self) {
+  fn refresh_dependencies(&mut self, root_dirs: Vec<Url>) {
     let resolver = self.resolver.as_graph_resolver();
     for doc in self.open_docs.values_mut() {
       if let Some(new_doc) = doc.maybe_with_new_resolver(resolver) {
         *doc = new_doc;
       }
     }
-    self.file_system_docs.lock().refresh_dependencies(resolver);
+
+    // update the file system document
+    let mut fs_docs = self.file_system_docs.lock();
+    let mut not_found_docs =
+      fs_docs.docs.keys().cloned().collect::<HashSet<_>>();
+
+    let mut pending_dirs = VecDeque::new();
+    for dir in root_dirs {
+      if let Ok(path) = dir.to_file_path() {
+        pending_dirs.push_back(path.to_path_buf());
+      }
+    }
+
+    while let Some(dir_path) = pending_dirs.pop_front() {
+      if let Some(dir_name) = dir_path.file_name() {
+        let dir_name = dir_name.to_string_lossy().to_lowercase();
+        if matches!(dir_name.as_str(), "node_modules" | ".git") {
+          continue;
+        }
+      }
+      let read_dir = match std::fs::read_dir(&dir_path) {
+        Ok(entry) => entry,
+        Err(_) => continue,
+      };
+
+      for entry in read_dir {
+        let entry = match entry {
+          Ok(entry) => entry,
+          Err(_) => continue,
+        };
+        let file_type = match entry.file_type() {
+          Ok(file_type) => file_type,
+          Err(_) => continue,
+        };
+        let new_path = dir_path.join(entry.file_name());
+        if file_type.is_dir() {
+          pending_dirs.push_back(new_path);
+        } else if file_type.is_file() {
+          let media_type: MediaType = (&new_path).into();
+          let is_diagnosable = match media_type {
```

Comment:
**nitpick:** I guess the match to if-no-match-return could be combined into something like:

```rs
match media_type {
   MediaType::Json
      | MediaType::Wasm
      | MediaType::SourceMap
      | MediaType::TsBuildInfo
      | MediaType::Unknown => {
      continue;
   },
   _ => {},
}
```

Reply:
The variable exists to describe that it's being skipped because it's not diagnosable and the explicit match for all entries is so a compiler error occurs when a new `MediaType` is added/removed/changed here so this code can be re-evaluated. For example, say `MediaType::Css` is added... there won't be an error here and someone might forget to update it. It was fairly common in the past that using `_ =>` would cause bugs.